common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -370,26 +391,19 @@ public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) throws IOExc
         new MergeStatuses(msg.shuffleId, new RoaringBitmap[0], new int[0], new long[0]);
     } else {
       Collection<AppShufflePartitionInfo> partitionsToFinalize = shufflePartitions.values();
-      int totalPartitions = partitionsToFinalize.size();
-      RoaringBitmap[] bitmaps = new RoaringBitmap[totalPartitions];
-      int[] reduceIds = new int[totalPartitions];
-      long[] sizes = new long[totalPartitions];
+      List<RoaringBitmap> bitmaps = new ArrayList<>(partitionsToFinalize.size());
+      List<Integer> reduceIds = new ArrayList<>(partitionsToFinalize.size());
+      List<Long> sizes = new ArrayList<>(partitionsToFinalize.size());
```

Comment:
We could alternatively still maintain the arrays to avoid array copying towards the end.
We can handle the null bitmaps on the driver side after it receives the MergeStatuses.

Reply:
I would prefer not sending null bitmaps. It seems awkward to do that. It is an API method so we also then have to document the API that it can send `nulls`. And what is the performance benefit? This happens once in finalize and the list will at most be few thousand elements big. Making the code optimize for this seems premature.

Reply:
I just thought that since we made this change in #30062, we might want to keep it if we can.
I don't have strong opinion on this.

Reply:
I think that was my mistake. I changed it overlooking the fact that there could be fewer partitions because of the exceptions when truncating data file or updating index/meta. 
I don't think it is better to have `null`s in the array that is returned though.